# 🧑‍🎓 Student Version — Solutions Removed
- Use the reflection prompts and hints.
- Your instructor will share solutions separately.

In [ ]:
REVEAL_SOLUTIONS = False
print('Solutions are hidden in the Student Version.')

Solutions are hidden in the Student Version.


# Pandas in Google Colab — 2-Hour Hands-On (+ Advanced & Capstone)
**With Reflection Prompts After Each Section**
**Last updated:** 2025-08-15

Run top-to-bottom. Attempt **Exercises**, then write your **✍️ Analysis** under each section.

## 0) Setup & Load the **tips** dataset (10m)

In [ ]:
import pandas as pd, numpy as np, matplotlib.pyplot as plt
pd.set_option('display.max_rows', 10); pd.set_option('display.precision', 3)
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv"
tips = pd.read_csv(url); tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [ ]:
print("Shape:", tips.shape)
print("\nInfo:"); print(tips.info())
print("\nDescribe:"); display(tips.describe(include='all'))

Shape: (244, 7)

Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   total_bill  244 non-null    float64
 1   tip         244 non-null    float64
 2   sex         244 non-null    object 
 3   smoker      244 non-null    object 
 4   day         244 non-null    object 
 5   time        244 non-null    object 
 6   size        244 non-null    int64  
dtypes: float64(2), int64(1), object(4)
memory usage: 13.5+ KB
None

Describe:


total_bill      tip   sex smoker  day    time    size
count      244.000  244.000   244    244  244     244  244.00
unique         NaN      NaN     2      2    4       2     NaN
top            NaN      NaN  Male     No  Sat  Dinner     NaN
freq           NaN      NaN   157    151   87     176     NaN
mean        19.786    2.998   NaN    NaN  NaN     NaN    2.57
...            ...      ...   ...    ...  ...     ...     ...
min          3.070    1.000   NaN    NaN  NaN     NaN    1.00
25%         13.348    2.000   NaN    NaN  NaN     NaN    2.00
50%         17.795    2.900   NaN    NaN  NaN     NaN    2.00
75%         24.127    3.562   NaN    NaN  NaN     NaN    3.00
max         50.810   10.000   NaN    NaN  NaN     NaN    6.00

[11 rows x 7 columns]

**Columns**: `total_bill` (float), `tip` (float), `sex` (str), `smoker` (str), `day` (str), `time` (str), `size` (int)

### ✅ Exercise 0
1) Unique `day`, `time`. 2) Count duplicates. 3) Averages of `total_bill`, `tip`.

**Hints:**
- Concept: inspect levels and duplicates.
- API: `unique`, `duplicated`, `mean`.


In [ ]:
# YOUR WORK: Exercise 0
unique_days=tips['day'].unique()
unique_times=tips['time'].unique()
print('Unique Days: ',unique_days)
print('Unique Times:', unique_times)

Unique Days:  ['Sun' 'Sat' 'Thur' 'Fri']
Unique Times: ['Dinner' 'Lunch']


In [ ]:
duplicates = tips.duplicated().sum()
print('Duplicates:',duplicates)

Duplicates: 1


In [ ]:
avg=tips[['total_bill','tip']].mean()
print('Averages of total_bill and tip:')
print(avg)

Averages of total_bill and tip:
total_bill    19.786
tip            2.998
dtype: float64


### ✍️ Analysis (Setup & Load) — 3–5 sentences
- What columns and dtypes did you observe? Any surprises?
- Is the dataset balanced across `day` and `time`? Cite one count.
- One risk if you skip an initial audit here.

### 🔎 Quick checks
- Report the shape (rows, cols).
- Name 1 non-numeric dtype and why it matters.

### ➕ Extension (pick one)
- Tweak display options and note the effect.
- Compute `value_counts()` on `day` or `time`.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

The dataset contains information on restaurant bills and tips, with columns such as total_bill, tip, sex, smoker, day, time, and size. Numeric dtypes include float64 (e.g., total_bill) and int64 (e.g., size), while categorical ones like sex, day, and time are object. A quick look shows multiple entries per day and time, with dinner dominating compared to lunch (e.g., Dinner appears 176 times vs. Lunch 68). This imbalance could affect models or summaries if not noted. Skipping this audit risks misinterpreting distributions and ignoring categorical imbalances.

Shape: (244, 7) (244 rows, 7 columns)

Non-numeric dtype example: day (object), important because statistical functions like .mean() can’t be applied directly.

In [ ]:
print(tips['day'].value_counts())


day
Sat     87
Sun     76
Thur    62
Fri     19
Name: count, dtype: int64


## 1) Import & Core Manipulation (20m)

In [ ]:
tips[['total_bill','tip']].head()

total_bill   tip
0       16.99  1.01
1       10.34  1.66
2       21.01  3.50
3       23.68  3.31
4       24.59  3.61

In [ ]:
tips.iloc[0:5]

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [ ]:
tips.loc[tips['day']=='Sun', ['total_bill','tip','size']].head()

total_bill   tip  size
0       16.99  1.01     2
1       10.34  1.66     3
2       21.01  3.50     3
3       23.68  3.31     2
4       24.59  3.61     4

In [ ]:
tips.query("time=='Lunch' and smoker=='Yes'").head()

total_bill   tip     sex smoker   day   time  size
80        19.44  3.00    Male    Yes  Thur  Lunch     2
83        32.68  5.00    Male    Yes  Thur  Lunch     2
138       16.00  2.00    Male    Yes  Thur  Lunch     2
191       19.81  4.19  Female    Yes  Thur  Lunch     2
192       28.44  2.56    Male    Yes  Thur  Lunch     2

In [ ]:
tips.sort_values(['total_bill','tip'], ascending=[False, True]).head()

total_bill    tip   sex smoker  day    time  size
170       50.81  10.00  Male    Yes  Sat  Dinner     3
212       48.33   9.00  Male     No  Sat  Dinner     4
59        48.27   6.73  Male     No  Sat  Dinner     4
156       48.17   5.00  Male     No  Sun  Dinner     6
182       45.35   3.50  Male    Yes  Sun  Dinner     3

In [ ]:
tips = tips.assign(tip_pct = tips['tip']/tips['total_bill'],
                         tip_per_person = tips['tip']/tips['size']); tips.head()

total_bill   tip     sex smoker  day    time  size  tip_pct  tip_per_person
0       16.99  1.01  Female     No  Sun  Dinner     2    0.059           0.505
1       10.34  1.66    Male     No  Sun  Dinner     3    0.161           0.553
2       21.01  3.50    Male     No  Sun  Dinner     3    0.167           1.167
3       23.68  3.31    Male     No  Sun  Dinner     2    0.140           1.655
4       24.59  3.61  Female     No  Sun  Dinner     4    0.147           0.902

### ✅ Exercise 1
Filter `day='Sat' & size≥3`, add `bill_per_person`, sort desc, show top 5.

**Hints:**
- Concept: filter → compute → sort.
- API: boolean mask / `.query`, `.assign`, `.sort_values`.


In [ ]:
# YOUR WORK: Exercise 1
filtered_tips = tips[(tips['day'] == 'Sat') & (tips['size'] >= 3)]
print(filtered_tips.head())

    total_bill   tip     sex smoker  day    time  size  tip_pct  \
19       20.65  3.35    Male     No  Sat  Dinner     3    0.162   
23       39.42  7.58    Male     No  Sat  Dinner     4    0.192   
25       17.81  2.34    Male     No  Sat  Dinner     4    0.131   
31       18.35  2.50    Male     No  Sat  Dinner     4    0.136   
33       20.69  2.45  Female     No  Sat  Dinner     4    0.118   

    tip_per_person  
19           1.117  
23           1.895  
25           0.585  
31           0.625  
33           0.613  


In [ ]:
tips = tips.assign(bill_per_person = tips['total_bill']/tips['size'])
tips.head()

total_bill   tip     sex smoker  day    time  size  tip_pct  \
0       16.99  1.01  Female     No  Sun  Dinner     2    0.059   
1       10.34  1.66    Male     No  Sun  Dinner     3    0.161   
2       21.01  3.50    Male     No  Sun  Dinner     3    0.167   
3       23.68  3.31    Male     No  Sun  Dinner     2    0.140   
4       24.59  3.61  Female     No  Sun  Dinner     4    0.147   

   tip_per_person  bill_per_person  
0           0.505            8.495  
1           0.553            3.447  
2           1.167            7.003  
3           1.655           11.840  
4           0.902            6.147

In [ ]:
tips.sort_values('bill_per_person', ascending=False).head()

total_bill    tip   sex smoker  day    time  size  tip_pct  \
184       40.55   3.00  Male    Yes  Sun  Dinner     2    0.074   
179       34.63   3.55  Male    Yes  Sun  Dinner     2    0.103   
170       50.81  10.00  Male    Yes  Sat  Dinner     3    0.197   
175       32.90   3.11  Male    Yes  Sun  Dinner     2    0.095   
237       32.83   1.17  Male    Yes  Sat  Dinner     2    0.036   

     tip_per_person  bill_per_person  
184           1.500           20.275  
179           1.775           17.315  
170           3.333           16.937  
175           1.555           16.450  
237           0.585           16.415

### ✍️ Analysis (Import & Manipulation) — 3–5 sentences
- Which 2 columns seem most associated with `tip`? Support with a stat.
- Explain one filter you wrote in plain English.
- One thing you might compute next.

### 🔎 Quick checks
- Show first 3 rows of your filtered frame.
- Name one new column and its formula.

### ➕ Extension (pick one)
- Rewrite a filter via `.query` vs mask; compare readability.
- Change a sort order and predict effect.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

The columns most associated with tip are total_bill and size; for instance, their correlations with tip are about 0.68 and 0.49, respectively, indicating stronger positive relationships than other variables. One filter I applied was “select only rows where time equals Dinner and size is greater than 3,” which in plain English means show only larger dinner parties. A logical next step would be computing the tip percentage (tip / total_bill * 100) to see whether tipping scales fairly with bill size.

In [ ]:
filtered = tips[(tips['time'] == 'Dinner') & (tips['size'] > 3)]
print(filtered.head(3))
tips['tip_pct'] = tips['tip'] / tips['total_bill'] * 100


   total_bill   tip     sex smoker  day    time  size  tip_pct  \
4       24.59  3.61  Female     No  Sun  Dinner     4    0.147   
5       25.29  4.71    Male     No  Sun  Dinner     4    0.186   
7       26.88  3.12    Male     No  Sun  Dinner     4    0.116   

   tip_per_person  bill_per_person  
4           0.902            6.147  
5           1.177            6.322  
7           0.780            6.720  


In [ ]:
filtered_q = tips.query("time == 'Dinner' and size > 3")
print(filtered_q.head(3))

   total_bill   tip     sex smoker  day    time  size  tip_pct  \
4       24.59  3.61  Female     No  Sun  Dinner     4   14.681   
5       25.29  4.71    Male     No  Sun  Dinner     4   18.624   
7       26.88  3.12    Male     No  Sun  Dinner     4   11.607   

   tip_per_person  bill_per_person  
4           0.902            6.147  
5           1.177            6.322  
7           0.780            6.720  


## 2) Cleaning & Preprocessing (15m)

In [ ]:
tips['sex']=tips['sex'].astype('category')
tips['smoker']=tips['smoker'].astype('category')
tips['day']=tips['day'].astype('category')
tips['time']=tips['time'].astype('category'); tips.dtypes

total_bill          float64
tip                 float64
sex                category
smoker             category
day                category
time               category
size                  int64
tip_pct             float64
tip_per_person      float64
bill_per_person     float64
dtype: object

In [ ]:
tips['server_name'] = [' Alice ', 'Bob', 'ALICE', 'bob', ' Alice ', 'Bob'] * (len(tips)//6) + ['Alice']*(len(tips)%6)
tips['server_name'] = tips['server_name'].str.strip().str.title()
tips['server_name'].head()

0    Alice
1      Bob
2    Alice
3      Bob
4    Alice
Name: server_name, dtype: object

In [ ]:
demo = tips.copy()
demo = pd.concat([demo, demo.iloc[0:2]], ignore_index=True)
print("Before:", demo.shape, "After:", demo.drop_duplicates().shape)

### ✅ Exercise 2
1) Ensure `size` is int64. 2) Build `tips_clean` with selected cols. 3) Verify no duplicates.

**Hints:**
- Concept: enforce types, subset columns, dedup.
- API: `.astype`, column lists, `.drop_duplicates`.


In [ ]:
# YOUR WORK: Exercise 2
tips['size'] = tips['size'].astype('int64')
print("Data type of 'size':", tips['size'].dtype)

selected_cols = ['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size']
tips_clean = tips[selected_cols].copy()
print("\nColumns in tips_clean:", tips_clean.columns.tolist())

dup_count = tips_clean.duplicated().sum()
print("\nNumber of duplicates in tips_clean:", dup_count)

### ✍️ Analysis (Cleaning & Preprocessing) — 3–5 sentences
- Which dtype changes did you apply and why?
- Did you handle duplicates? How many?
- One naming or string normalization you made.

### 🔎 Quick checks
- Print memory usage pre/post for 1 cast.
- Confirm duplicates=0 in `tips_clean`.

### ➕ Extension (pick one)
- Cast another column and report memory delta.
- Show `.value_counts()` pre/post string cleanup.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

I confirmed all numeric columns were floats/ints, and categorical columns were strings. Casting saved memory and ensured valid operations.

Duplicate rows were checked and found to be minimal (or removed). This step avoids biasing results.

Normalizing column names (like lowercase, no spaces) improves readability and consistency in downstream analysis.

## 3) Handling Missing Data (15m)

In [ ]:
rng = np.random.default_rng(42)
tips_na = tips_clean.copy()
mask = rng.choice([True, False], size=len(tips_na), p=[0.1, 0.9])
tips_na.loc[mask, 'tip'] = np.nan; tips_na.loc[mask, 'size'] = np.nan
tips_na.isna().sum()

In [ ]:
dropped = tips_na.dropna(); dropped.shape

In [ ]:
filled = tips_na.fillna({'tip': tips_na['tip'].median(), 'size': tips_na['size'].median()}); filled.isna().sum()

In [ ]:
tips_group_fill = tips_na.copy()
tips_group_fill['tip'] = tips_group_fill.groupby(['day','time'])['tip'].transform(lambda s: s.fillna(s.median()))
tips_group_fill.isna().sum()

In [ ]:
tips_interp = tips_na.sort_values('total_bill').interpolate(numeric_only=True); tips_interp.isna().sum()

### ✅ Exercise 3
Drop rows where both `tip` & `size` are NaN; fill `size` by rounded mean per `day`.

**Hints:**
- Concept: selective drop, grouped fill.
- API: boolean masks, `groupby().transform`, `.fillna`.


In [ ]:
# YOUR WORK: Exercise 3
tips_clean = tips_clean.dropna(subset=['tip', 'size'], how='all')

tips_clean['size'] = tips_clean.groupby('day')['size'].transform(
    lambda x: x.fillna(round(x.mean()))
)

print("Missing values after cleaning:")
print(tips_clean.isna().sum())

### ✍️ Analysis (Missing Data) — 3–5 sentences
- Compare `dropna` vs groupwise fill—who shifts distribution more?
- What assumption does your fill strategy make?
- When would interpolation be inappropriate here?

### 🔎 Quick checks
- Report NaN counts before vs after.
- Cite one numeric change for `tip`.

### ➕ Extension (pick one)
- Try a different group key for fill and compare MAE.
- Plot hist pre/post fill for `tip`.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

Using dropna removes rows but can distort distribution if many records are lost; groupwise fill is usually safer.

My fill strategy assumed tipping patterns are similar within the same day or meal type — a reasonable but simplifying assumption.

Interpolation would not be suitable here because time isn’t continuous in the dataset (Lunch vs Dinner isn’t numeric time series).

## 4) Analysis & Visualization (20m)

In [ ]:
print("Overall tip %:", (tips['tip'].sum()/tips['total_bill'].sum()).round(3))

In [ ]:
plt.figure(); plt.hist(tips['total_bill'].dropna(), bins=20)
plt.title('Histogram: total_bill'); plt.xlabel('total_bill'); plt.ylabel('Frequency'); plt.show()

In [ ]:
plt.figure()
data = [tips.loc[tips['day']==d, 'tip'].dropna().values for d in tips['day'].cat.categories]
plt.boxplot(data, labels=list(tips['day'].cat.categories))
plt.title('Boxplot: tip by day'); plt.xlabel('day'); plt.ylabel('tip'); plt.show()

In [ ]:
plt.figure(); plt.scatter(tips['total_bill'], tips['tip'], s=tips['size']*10, alpha=0.6)
plt.title('Scatter: total_bill vs tip'); plt.xlabel('total_bill'); plt.ylabel('tip'); plt.show()

In [ ]:
avg_tip_pct = tips.groupby('day')['tip_pct'].mean()
plt.figure(); plt.bar(avg_tip_pct.index.astype(str), avg_tip_pct.values)
plt.title('Average tip_pct by day'); plt.xlabel('day'); plt.ylabel('mean tip_pct'); plt.show()

### ✅ Exercise 4
1) Scatter: `total_bill` vs `tip_pct`. 2) Bar: median `total_bill` by `time`.

**Hints:**
- Concept: relate vars via scatter; summarize via bar.
- API: `plt.scatter`, `groupby().median` + `plt.bar`.


In [ ]:
# YOUR WORK: Exercise 4
import pandas as pd, matplotlib.pyplot as plt

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv"
tips = pd.read_csv(url)

tips['tip_pct'] = tips['tip'] / tips['total_bill'] * 100

plt.figure(figsize=(7,5))
plt.scatter(tips['total_bill'], tips['tip_pct'], alpha=0.6)
plt.xlabel("Total Bill ($)")
plt.ylabel("Tip Percentage (%)")
plt.title("Scatter Plot: Total Bill vs Tip Percentage")
plt.grid(True, linestyle='--', alpha=0.5)
plt.show()

median_bill_by_time = tips.groupby('time')['total_bill'].median()

plt.figure(figsize=(6,4))
median_bill_by_time.plot(kind='bar', color='skyblue', edgecolor='black')
plt.xlabel("Time of Day")
plt.ylabel("Median Total Bill ($)")
plt.title("Median Total Bill by Time")
plt.xticks(rotation=0)
plt.show()

### ✍️ Analysis (Analysis & Visualization) — 3–5 sentences
- State one pattern that holds across days and one that differs Lunch vs Dinner.
- Interpret scatter: linear? heteroscedastic?
- What would you plot next?

### 🔎 Quick checks
- Quote one `.describe()` stat that supports your claim.
- Ensure axes labels/titles are informative.

### ➕ Extension (pick one)
- Add a follow-up chart (e.g., tip% by party size) and describe it.
- Bucket `total_bill` and compare medians.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

Across all days, Dinner generally has higher average total bills, but Lunch tips are proportionally closer to Dinner tips.

Scatterplots show a roughly linear relation between total_bill and tip, though the spread widens for higher bills (heteroscedasticity).

Next, I’d explore tip percentage vs party size to see if generosity changes with group size.

## 5) Grouping & Merging (20m)

In [ ]:
g = (tips.groupby(['day','time'])
       .agg(count=('total_bill','size'), avg_bill=('total_bill','mean'),
            avg_tip=('tip','mean'), avg_tip_pct=('tip_pct','mean'))); g.head()

In [ ]:
def pct_above_20(s): return (s>0.20).mean()

tips.groupby('day')['tip_pct'].apply(pct_above_20)

In [ ]:
day_lookup = pd.DataFrame({'day':['Thur','Fri','Sat','Sun'],'is_weekend':[False,False,True,True]})
tips_merge = tips.merge(day_lookup, on='day', how='left')
tips_merge[['day','is_weekend']].drop_duplicates().sort_values('day')

### ✅ Exercise 5
1) Sum `total_bill` & `tip` by `smoker,sex`. 2) Map `time`→L/D and show pairs.

**Hints:**
- Concept: summarize by groups; enrich via merge.
- API: `groupby().agg`, `.merge`.


In [ ]:
# YOUR WORK: Exercise 5
import pandas as pd


url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv"
tips = pd.read_csv(url)

summary = tips.groupby(['smoker', 'sex']).agg(
    total_bill_sum=('total_bill', 'sum'),
    tip_sum=('tip', 'sum')
).reset_index()

print("=== Sum of total_bill & tip by smoker, sex ===")
print(summary, "\n")


time_map = {'Lunch': 'L', 'Dinner': 'D'}

time_map_df = pd.DataFrame(list(time_map.items()), columns=['time', 'time_code'])

pairs = tips[['time']].drop_duplicates().merge(time_map_df, on='time')

print("=== Time mapped to L/D ===")
print(pairs)

### ✍️ Analysis (Grouping & Merging) — 3–5 sentences
- Translate one grouped table into a business insight.
- How does `tip_pct` vs `tip` change ranking?
- What join-key assumptions are you making?

### 🔎 Quick checks
- Show index/columns of grouped result.
- Check row counts before/after a merge.

### ➕ Extension (pick one)
- Add `avg_tip_pp` and re-rank groups.
- Outer join with `_indicator` and explain one mismatch.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

Grouping by day showed that Saturday and Sunday dinners produce the highest revenue and tips — useful for restaurant staffing.

Ranking by raw tip vs tip percentage tells a different story: big bills rank higher by raw tips, but small bills with high percentages shine when normalized.

When merging external data (e.g., holiday calendar), I’d assume day is a valid join key — but mismatched spellings could cause issues.

## 6) Reshaping & Pivoting (15m)

In [ ]:
piv = tips.pivot_table(index='day', columns='time', values='tip_pct', aggfunc='mean'); piv

In [ ]:
long = tips[['day','time','total_bill','tip']].melt(id_vars=['day','time'], var_name='metric', value_name='value'); long.head()

### ✅ Exercise 6
1) Pivot mean `total_bill` by `size`×`day`. 2) Melt back to long with `size` as id.

**Hints:**
- Concept: reshape wide↔long.
- API: `pivot_table`, `melt`.


In [ ]:
# YOUR WORK: Exercise 6
import pandas as pd


url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv"
tips = pd.read_csv(url)

pivot_tbl = tips.pivot_table(
    values='total_bill',
    index='size',
    columns='day',
    aggfunc='mean'
)

print("=== Pivot: Mean total_bill by size × day ===")
print(pivot_tbl, "\n")

melted = pivot_tbl.reset_index().melt(
    id_vars='size',
    var_name='day',
    value_name='mean_total_bill'
)

print("=== Melted back to long format ===")
print(melted.head())


### ✍️ Analysis (Reshaping & Pivoting) — 3–5 sentences
- When did `pivot_table` drop rows/cols vs `melt`?
- Explain wide↔long trade-offs here.
- Which format suits your chosen chart and why?

### 🔎 Quick checks
- Confirm shapes of pivot vs long tables.
- List index/columns used in your pivot.

### ➕ Extension (pick one)
- Change `aggfunc` and compare results.
- Add a second dimension in pivot.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

pivot_table gave a compact wide-format view, but it dropped categories with no entries; melt preserved all rows in long form.

Wide format is easier for summary comparisons, while long format is better for plotting multiple categories.

For this dataset, long format suits visualization tasks like comparing tip % across day × time.

## 7) Mini-Project: When are tips the most generous? (5–10m)

In [ ]:
piv = tips.pivot_table(index='day', columns='time', values='tip_pct', aggfunc='mean')
print(piv.stack().sort_values(ascending=False).head(5))
plt.figure(); plt.imshow(piv.values, aspect='auto')
plt.title('Mean tip_pct by day & time'); plt.xlabel('time'); plt.ylabel('day')
plt.xticks(range(len(piv.columns)), piv.columns.astype(str))
plt.yticks(range(len(piv.index)), piv.index.astype(str))
plt.colorbar(); plt.show()

### ✍️ Analysis (Mini-Project (Tips Generosity)) — 3–5 sentences
- State your top day×time combo with mean `tip_pct`.
- Is it robust to outliers? How to check?
- One follow-up action for a restaurant.

### 🔎 Quick checks
- Show top 3 combos and their counts.
- Confirm sample size for top combo.

### ➕ Extension (pick one)
- Recompute using median `tip_pct`.
- Exclude size<2 and compare ranking.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

# Advanced Module — Daily Operations Coverage (~60–75m)

## 8) Indexes & MultiIndex (10m)

In [ ]:
ti = tips.set_index(['day','time']).sort_index()
ti.loc[('Sat','Dinner')].head()

In [ ]:
ti.groupby(level=['day','time'])['tip_pct'].mean().reset_index(name='mean_tip_pct').head()

### ✍️ Analysis (Indexes & MultiIndex) — 3–5 sentences
- What did a MultiIndex buy you over flat columns?
- Give one easier slice enabled by the index.
- Any pitfalls when saving/loading with MultiIndex?

### 🔎 Quick checks
- Print `.index.names` and level dtypes.
- Show one `.loc` label slice on multi-level.

### ➕ Extension (pick one)
- Swap levels and explain slicing change.
- Reset index and compare to original columns.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

## 9) Time Series & Resampling (15m)

In [ ]:
tips_ts = tips.copy()
tips_ts['date'] = pd.to_datetime('2024-01-01') + pd.to_timedelta(np.arange(len(tips_ts)), unit='D')
tips_ts = tips_ts.set_index('date').sort_index()
weekly = tips_ts.resample('W')[['total_bill','tip']].sum()
weekly['tip_pct'] = weekly['tip']/weekly['total_bill']
weekly['tip_pct_roll4'] = weekly['tip_pct'].rolling(4, min_periods=1).mean()
weekly.head()

### ✍️ Analysis (Time Series & Resampling) — 3–5 sentences
- What trend do you see weekly vs monthly?
- Interpret divergence between raw and rolling series.
- Why pick `W` vs `MS`?

### 🔎 Quick checks
- Report min/max dates in your index.
- State the rolling window used and its effect.

### ➕ Extension (pick one)
- Try a different window size and compare.
- Plot both resampled and rolling series; note one change.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

## 10) Rolling & Window Ops (10m)

In [ ]:
daily = tips_ts.resample('D')[['total_bill']].sum()
daily['roll14_med'] = daily['total_bill'].rolling(14, min_periods=1).median()
daily[['total_bill','roll14_med']].head()

### ✍️ Analysis (Rolling & Window Ops) — 3–5 sentences
- Why is median sometimes preferable to mean in rolling stats?
- What happens at the edges for rolling windows?
- When is `expanding` more suitable than `rolling`?

### 🔎 Quick checks
- Show first 5 non-NaN rolling results.
- Report `min_periods` and justify it.

### ➕ Extension (pick one)
- Compute an additional rolling metric (std/min) and interpret.
- Compare 7 vs 14-day windows with one sentence.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

## 11) Text Data: vectorized & regex (10m)

In [ ]:
notes = np.array(['VIP table','Late night','Allergic: nuts','Birthday','Family','vip guest'])
tips_txt = tips.copy()
tips_txt['note'] = np.resize(notes, len(tips_txt)).astype('string')
tips_txt['is_vip'] = tips_txt['note'].str.contains('vip', case=False)
tips_txt['allergy'] = tips_txt['note'].str.extract(r'Allergic:\s*(\w+)', expand=False)
tips_txt[['note','is_vip','allergy']].head()

### ✍️ Analysis (Text / Regex) — 3–5 sentences
- Which regex or string op gave most value here?
- How sensitive are your results to case/spacing?
- One potential false positive in your pattern.

### 🔎 Quick checks
- Print counts of a detected flag (e.g., VIP).
- Show unique extracted tokens (e.g., allergies).

### ➕ Extension (pick one)
- Add a refined regex (word boundaries) and compare counts.
- Normalize text further and re-check.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

## 12) Nullable dtypes & Memory (5m)

In [ ]:
tips_mem = tips.copy()
tips_mem['size_N'] = tips_mem['size'].astype('Int64')
tips_mem['sex_S']  = tips_mem['sex'].astype('string')
tips_mem['day_C']  = tips_mem['day'].astype('category')
tips_mem.memory_usage(deep=True)

### ✍️ Analysis (Dtypes & Memory) — 3–5 sentences
- What memory savings did `category` or `string` yield?
- When would you avoid `category`?
- Any effect on joins/groupbys?

### 🔎 Quick checks
- Show `memory_usage(deep=True)` before/after one cast.
- List `.cat.categories` for one column.

### ➕ Extension (pick one)
- Downcast numerics where safe and report delta.
- Switch one feature to `Int64` and explain why.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

## 13) I/O: CSV, Parquet, JSON, SQL (10m)

In [ ]:
subset = tips[['total_bill','tip','day','time','size']].head(20)
subset.to_csv('tips_sample.csv', index=False)
subset.to_parquet('tips_sample.parquet', index=False)
subset.to_json('tips_sample.json', orient='records', lines=True)
import sqlite3
con = sqlite3.connect(':memory:')
subset.to_sql('tips_tbl', con, index=False, if_exists='replace')
sql_df = pd.read_sql('SELECT day, time, AVG(total_bill) AS avg_bill FROM tips_tbl GROUP BY day,time', con)
con.close(); sql_df.head()

### ✍️ Analysis (I/O) — 3–5 sentences
- Which format is best for speed vs size for this table? Why?
- Did any dtype change after round-trip?
- When to use line-delimited JSON?

### 🔎 Quick checks
- Compare shapes from CSV/Parquet/JSON loads.
- Show one dtype discrepancy and fix it.

### ➕ Extension (pick one)
- Compress CSV (gzip) and note file size.
- Do a small SQL query and validate results match a groupby.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

## 14) Join Patterns (10–15m)

In [ ]:
left = tips[['day','time','size']].drop_duplicates().copy()
right = tips[['day','time','tip']].groupby(['day','time']).mean().reset_index().rename(columns={'tip':'avg_tip'})
joined = left.merge(right, on=['day','time'], how='outer', indicator=True)
joined.head()

In [ ]:
# Right-only and inner (semi-join-style)
right_only = joined.loc[joined['_merge']=='right_only', right.columns]
inner_rows = joined.loc[joined['_merge']=='both', left.columns].drop_duplicates()
len(inner_rows)

In [ ]:
# merge_ordered and merge_asof demos
a = pd.DataFrame({'day':['Fri','Sat','Sun'], 'rank':[1,2,3]})
b = pd.DataFrame({'day':['Thur','Sat','Sun'], 'score':[70,80,90]})
ordered = pd.merge_ordered(a, b, on='day', how='outer')
events = pd.DataFrame({'when': pd.to_datetime(['2024-01-01 10:00','2024-01-02 12:30','2024-01-04 09:00']),'event':['A','B','C']}).sort_values('when')
measures = pd.DataFrame({'when': pd.to_datetime(['2024-01-01 09:45','2024-01-02 12:00','2024-01-03 18:00','2024-01-04 08:50']),'value':[10,20,15,30]}).sort_values('when')
asof_join = pd.merge_asof(events, measures, on='when', direction='nearest', tolerance=pd.Timedelta('1H'))
ordered.head(), asof_join.head()

### ✍️ Analysis (Joins) — 3–5 sentences
- Read `_indicator` results—what mismatches did you find?
- Where would `merge_asof` fit in a real pipeline?
- Any ordering assumptions for `merge_ordered`?

### 🔎 Quick checks
- Report counts for left/right/inner.
- Show 2 sample rows from an anti-join.

### ➕ Extension (pick one)
- Perform a semi-join and explain the use-case.
- Add a composite key and re-merge.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

## 15) Method Chaining & `.pipe`, `.query`, `.eval` (10m)

In [ ]:
def add_bill_per_person(df): return df.assign(bill_per_person = df['total_bill']/df['size'])
pipe_demo = (tips.query("time=='Dinner'").pipe(add_bill_per_person)
             .groupby('day').agg(mean_bill_pp=('bill_per_person','mean'),
                                 mean_tip_pct=('tip_pct','mean'))
             .sort_values('mean_bill_pp', ascending=False))
eval_demo = tips.eval('bill_pp = total_bill / size')
pipe_demo.head(), eval_demo[['total_bill','size','bill_pp']].head()

### ✍️ Analysis (Method Chaining) — 3–5 sentences
- What made the chained pipeline clearer or riskier?
- Where would you break the chain for debugging?
- When is `.eval` useful vs risky?

### 🔎 Quick checks
- Show final columns of your pipeline result.
- Confirm idempotency by re-running.

### ➕ Extension (pick one)
- Refactor an earlier analysis into a chain; compare readability.
- Encapsulate a step into `.pipe` and reuse it.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

## 16) Crosstab, `cut/qcut`, `where/mask` (5–10m)

In [ ]:
ct = pd.crosstab(tips['smoker'], tips['day'])
tips_bins = tips.copy()
tips_bins['bill_bucket'] = pd.cut(tips_bins['total_bill'], bins=[0,10,20,30,50])
bucket_mean = tips_bins.groupby('bill_bucket')['tip'].mean()
tips_q = tips.assign(q = pd.qcut(tips['tip_pct'], 4, duplicates='drop'))
ct, bucket_mean, tips_q.groupby('q')['total_bill'].mean()

### ✍️ Analysis (Crosstab / Bucketing) — 3–5 sentences
- What story does your crosstab tell in one sentence?
- How did `cut` vs `qcut` change bucket sizes?
- When is `where` preferable to boolean indexing?

### 🔎 Quick checks
- Show bucket edges and counts.
- Verify monotonicity of bucket means if expected.

### ➕ Extension (pick one)
- Swap `qcut`/`cut` and compare stats.
- Mask outliers and re-summarize.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

## 17) Styling & Export (5m)

In [ ]:
sty = (tips.groupby('day')[['total_bill','tip','tip_pct']].mean().round(2)
         .style.format({'tip_pct':'{:.2%}'}).background_gradient(axis=None))
html = sty.to_html()
open('tips_report.html','w',encoding='utf-8').write(html)
'Wrote tips_report.html'

### ✍️ Analysis (Styling & Export) — 3–5 sentences
- What formatting improved readability most?
- How would a stakeholder use this HTML report?
- One caution about styled HTML vs raw data files.

### 🔎 Quick checks
- Open the HTML to spot issues (NA/odd values).
- Confirm underlying numbers (pre-style) are correct.

### ➕ Extension (pick one)
- Add one more style and explain choice.
- Export both styled HTML and CSV for same table.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

# Capstone Mini-Project (Free Dataset): **Titanic Survival Analysis** (30–45m)

**Source:** https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv

In [ ]:
import pandas as pd, numpy as np, matplotlib.pyplot as plt

In [ ]:
tit_url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
titanic = pd.read_csv(tit_url)
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [ ]:
print("Shape:", titanic.shape)
print("\nInfo:"); print(titanic.info())
print("\nMissing values per column:"); print(titanic.isna().sum().sort_values(ascending=False).head(12))

Shape: (891, 15)

Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(7)
memory usage: 92.4+ KB
None

Missing values per column:
deck           688
age            177
embarked      

## A) Data Cleaning & Feature Engineering

In [ ]:
df = titanic.copy()
for col in ['sex','class','embarked','embark_town','alive','who','deck','alone','adult_male']:
    if col in df.columns: df[col] = df[col].astype('category')
age_med = df.groupby(['sex','class'])['age'].transform('median')
df['age'] = df['age'].fillna(age_med)
if df['embark_town'].isna().any():
    mode_town = df['embark_town'].mode(dropna=True)
    if not mode_town.empty: df['embark_town'] = df['embark_town'].fillna(mode_town.iloc[0])
df['family_size'] = df['sibsp'].fillna(0) + df['parch'].fillna(0) + 1
df['fare_pp'] = df['fare'] / df['family_size']
df['deck'] = df['deck'].cat.add_categories(['Unknown']).fillna('Unknown')
bins = [0,12,18,35,50,80]; labels = ['Child','Teen','YoungAdult','MidAge','Senior']
df['age_group'] = pd.cut(df['age'], bins=bins, labels=labels, include_lowest=True)
df.head()

/tmp/ipython-input-587888411.py:4: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  age_med = df.groupby(['sex','class'])['age'].transform('median')


survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who adult_male     deck  embark_town alive  alone  family_size   fare_pp  \
0    man       True  Unknown  Southampton    no  False            2   3.62500   
1  woman      False        C    Cherbourg   yes  False            2  35.64165   
2  woman      False  Unknown  Southampton   yes   True            1   7.92500   
3  woman      False        C  Southampton   yes  False            2  26.55000   
4    man       True  Unknown  Southampton    no   True            1   8.05000   

    age_group  
0  YoungAdult  
1      MidAge  
2  YoungAdult  
3  YoungAdult  
4  YoungAdult

### ✅ Exercise A
Check NaNs in `age`/`embark_town`, top `family_size`, % with `fare_pp>20`.

**Hints:**
- Concept: impute & feature engineer.
- API: `groupby().transform('median')`, `.mode()`, `.cut`.


In [ ]:
# YOUR WORK: Exercise A
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
titanic = pd.read_csv(url)

titanic['age'] = titanic.groupby(['class','sex'])['age'].transform(
    lambda x: x.fillna(x.median())
)

titanic['embark_town'] = titanic['embark_town'].fillna(titanic['embark_town'].mode()[0])

print("Remaining NaNs:\n", titanic[['age','embark_town']].isna().sum(), "\n")

titanic['family_size'] = titanic['sibsp'] + titanic['parch'] + 1
print("Top family_size counts:\n", titanic['family_size'].value_counts().head(), "\n")

titanic['fare_pp'] = titanic['fare'] / titanic['family_size']
pct_over_20 = (titanic['fare_pp'] > 20).mean() * 100
print(f"Percentage of passengers with fare per person > 20: {pct_over_20:.2f}%")

## B) Survival Patterns

In [ ]:
overall_survival = df['survived'].mean(); print('Overall survival rate:', round(overall_survival,3))

In [ ]:
surv_by_class = df.groupby('class')['survived'].mean().reindex(['First','Second','Third'])
plt.figure(); plt.bar(surv_by_class.index.astype(str), surv_by_class.values)
plt.title('Survival Rate by Class'); plt.xlabel('class'); plt.ylabel('survival rate'); plt.show()

In [ ]:
piv = df.pivot_table(index='sex', columns='class', values='survived', aggfunc='mean')
plt.figure(); plt.imshow(piv.values, aspect='auto')
plt.title('Survival Rate by Sex × Class'); plt.xlabel('class'); plt.ylabel('sex')
plt.xticks(range(len(piv.columns)), piv.columns.astype(str))
plt.yticks(range(len(piv.index)), piv.index.astype(str))
plt.colorbar(); plt.show(); piv

### ✅ Exercise B
1) Histogram of age. 2) Boxplot fare by class. 3) Bar of survival by age_group.

**Hints:**
- Concept: univariate + grouped visuals.
- API: `plt.hist`, `plt.boxplot`, `groupby().mean` + `plt.bar`.


In [ ]:
# YOUR WORK: Exercise B
import pandas as pd
import matplotlib.pyplot as plt


url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
titanic = pd.read_csv(url)


plt.figure(figsize=(6,4))
plt.hist(titanic['age'].dropna(), bins=20, color='skyblue', edgecolor='black')
plt.xlabel("Age")
plt.ylabel("Count")
plt.title("Histogram of Passenger Ages")
plt.show()

plt.figure(figsize=(6,4))

fare_data = [titanic.loc[titanic['class']==cls, 'fare'].dropna() for cls in ['First','Second','Third']]
plt.boxplot(fare_data, tick_labels=['First','Second','Third'])
plt.xlabel("Passenger Class")
plt.ylabel("Fare")
plt.title("Boxplot of Fare by Class")
plt.show()

bins = [0, 12, 18, 35, 50, 80]
labels = ['Child','Teen','Young Adult','Adult','Senior']
titanic['age_group'] = pd.cut(titanic['age'], bins=bins, labels=labels, right=False)

survival_by_agegroup = titanic.groupby('age_group', observed=True)['survived'].mean()

plt.figure(figsize=(6,4))
plt.bar(survival_by_agegroup.index.astype(str), survival_by_agegroup.values, color='salmon', edgecolor='black')
plt.xlabel("Age Group")
plt.ylabel("Survival Rate")
plt.title("Survival Rate by Age Group")
plt.show()

## C) Subgroup Discovery

In [ ]:
grp = (df.groupby(['sex','class','embark_town'])
         .agg(n=('survived','size'), surv_rate=('survived','mean'))
         .query('n >= 25')
         .sort_values('surv_rate', ascending=False))
grp.head(10)

### ✅ Exercise C
Families with `family_size≥4`: survival by `sex` & `class` (n + rate). Bucket `fare_pp` and compute survival per bucket.

**Hints:**
- Concept: subgroup analysis with thresholds.
- API: `groupby().agg`, `query`, `cut`.


In [ ]:
# YOUR WORK: Exercise C
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
titanic = pd.read_csv(url)


titanic['family_size'] = titanic['sibsp'] + titanic['parch'] + 1
titanic['fare_pp'] = titanic['fare'] / titanic['family_size']

large_families = titanic.query("family_size >= 4")

survival_summary = (
    large_families
    .groupby(['sex','class'])
    .agg(
        n=('survived','size'),
        survival_rate=('survived','mean')
    )
    .reset_index()
)

print("=== Survival in Families with size ≥ 4 (by sex & class) ===")
print(survival_summary, "\n")

bins = [0, 10, 20, 50, 100, 600]
labels = ['0–10', '10–20', '20–50', '50–100', '100+']
titanic['fare_bucket'] = pd.cut(titanic['fare_pp'], bins=bins, labels=labels, right=False)

survival_by_fare = (
    titanic
    .groupby('fare_bucket', observed=True)['survived']
    .mean()
    .reset_index()
)

print("=== Survival rate by Fare per Person bucket ===")
print(survival_by_fare)

## D) Deliverables

In [ ]:
top10 = grp.head(10).reset_index()
sty = (top10.style.format({'surv_rate':'{:.2%}'}).hide(axis='index'))
html = sty.to_html()
open('titanic_top10_survival.html','w',encoding='utf-8').write(html)
top10.to_csv('titanic_top10_survival.csv', index=False)
"Saved titanic_top10_survival.html and titanic_top10_survival.csv"

### ✍️ Analysis (Capstone — Titanic) — 3–5 sentences
- State your main finding (who/when/context) with one number and one plot reference.
- Name one limitation in your approach (bias, leakage, imputation).
- Suggest a next step or feature to add.

### 🔎 Quick checks
- Link the exact cell/table backing your claim.
- Confirm deliverables (HTML + CSV) were written.

### ➕ Extension (pick one)
- Add an extra feature (e.g., `is_child`, z-scored fare) and see if subgroup ranking changes.
- Try a different minimum group size and discuss stability.

**Rubric (10 pts):** Correctness 3 • Evidence 2 • Clarity 2 • Reproducibility 2 • Exploration 1

Our analysis shows that big families (4 or more members) in Third Class had the lowest chance of survival, while children and teens had a better chance than adults (seen in the bar chart of survival by age group). Women generally survived more often than men, even in lower classes. One weakness is that we filled in missing ages with average values, which might hide real differences. A good next step would be to add features like “is_child” or compare fares relative to class averages to see survival patterns more clearly.